gardenlet: invalid memory address or nil pointer dereference

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
-->
/area quality ops-productivity
/kind bug

**What happened**:
```
% k -n garden logs gardenlet-746c5fb7d9-qz84x

panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x1edea2d]

goroutine 27363 [running]:
github.com/gardener/gardener/pkg/gardenlet/controller/shoot/shoot.(*Reconciler).runReconcileShootFlow.func13({0x2a0d500, 0xc0ba720210})
	/go/src/github.com/gardener/gardener/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go:391 +0x4d
github.com/gardener/gardener/pkg/utils/flow.TaskFn.RetryUntilTimeout.func1.1({0x2a0d500?, 0xc0ba720210?})
	/go/src/github.com/gardener/gardener/pkg/utils/flow/taskfn.go:74 +0x27
github.com/gardener/gardener/pkg/utils/retry.UntilFor({0x2a0d500, 0xc0ba720210}, 0x22051c0?, {0x29fcf80?, 0xc0b3cc9400?}, 0xc0b3cc93f0)
	/go/src/github.com/gardener/gardener/pkg/utils/retry/retry.go:142 +0x59
github.com/gardener/gardener/pkg/utils/retry.(*ops).Until(0xc00072b500, {0x2a0d500, 0xc0ba720210}, 0x22051c0?, 0x8bb2c97001?)
	/go/src/github.com/gardener/gardener/pkg/utils/retry/retry.go:186 +0x75
github.com/gardener/gardener/pkg/utils/flow.TaskFn.RetryUntilTimeout.func1({0x2a0d538?, 0xc089b7e300?})
	/go/src/github.com/gardener/gardener/pkg/utils/flow/taskfn.go:73 +0xc3
github.com/gardener/gardener/pkg/utils/flow.(*execution).runNode.func1()
	/go/src/github.com/gardener/gardener/pkg/utils/flow/flow.go:212 +0x219
created by github.com/gardener/gardener/pkg/utils/flow.(*execution).runNode
	/go/src/github.com/gardener/gardener/pkg/utils/flow/flow.go:208 +0x20f
```

The nil pointer comes from https://github.com/gardener/gardener/blob/0f62246189813cc25088fd65a23b2d6d1a5262ce/pkg/gardenlet/controller/shoot/shoot/reconciler_reconcile.go#L391. `o.ShootClientSet` is nil for a Shoot with enabled hibernation:
https://github.com/gardener/gardener/blob/48bd154e4b3f2644d38f8a7a328976f85b446d36/pkg/operation/operation.go#L307-L346

**What you expected to happen**:
No such panic.

**How to reproduce it (as minimally and precisely as possible)**:
Trigger ETCD encryption key rotation and hibernation at the same time.
Or maybe even trigger etcd encryption key rotation for hibernated cluster.

**Anything else we need to know?**:

**Environment**:

- Gardener version: v1.73.1
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

--- a/pkg/apis/core/helper/helpers.go
+++ b/pkg/apis/core/helper/helpers.go
@@ -337,7 +337,16 @@ func GetShootAuditPolicyConfigMapRef(apiServerConfig *core.KubeAPIServerConfig) 
 
 // HibernationIsEnabled checks if the given shoot's desired state is hibernated.
 func HibernationIsEnabled(shoot *core.Shoot) bool {
-	return shoot.Spec.Hibernation != nil && shoot.Spec.Hibernation.Enabled != nil && *shoot.Spec.Hibernation.Enabled
+	return shoot.Spec.Hibernation != nil && pointer.BoolDeref(shoot.Spec.Hibernation.Enabled, false)
+}
+
+// HibernationIsEnabled checks if the given shoot is in hibernation or is waking up.
+func IsShootInHibernation(shoot *core.Shoot) bool {
+	if shoot.Spec.Hibernation != nil && shoot.Spec.Hibernation.Enabled != nil {
+		return *shoot.Spec.Hibernation.Enabled || shoot.Status.IsHibernated
+	}
+
+	return shoot.Status.IsHibernated
 }
 
 // SystemComponentsAllowed checks if the given worker allows system components to be scheduled onto it

--- a/pkg/apis/core/helper/helpers.go
+++ b/pkg/apis/core/helper/helpers.go
@@ -340,7 +340,7 @@ func HibernationIsEnabled(shoot *core.Shoot) bool {
 	return shoot.Spec.Hibernation != nil && pointer.BoolDeref(shoot.Spec.Hibernation.Enabled, false)
 }
 
-// HibernationIsEnabled checks if the given shoot is in hibernation or is waking up.
+// IsShootInHibernation checks if the given shoot is in hibernation or is waking up.
 func IsShootInHibernation(shoot *core.Shoot) bool {
 	if shoot.Spec.Hibernation != nil && shoot.Spec.Hibernation.Enabled != nil {
 		return *shoot.Spec.Hibernation.Enabled || shoot.Status.IsHibernated

--- a/pkg/apis/core/helper/helpers_test.go
+++ b/pkg/apis/core/helper/helpers_test.go
@@ -488,6 +488,34 @@ var _ = Describe("helper", func() {
 		}, true),
 	)
 
+	DescribeTable("#IsShootInHibernation",
+		func(shoot *core.Shoot, hibernated bool) {
+			Expect(IsShootInHibernation(shoot)).To(Equal(hibernated))
+		},
+		Entry("no hibernation section and status.isHibernated is false", &core.Shoot{}, false),
+		Entry("no hibernation section and status.isHibernated is true", &core.Shoot{
+			Status: core.ShootStatus{IsHibernated: true},
+		}, true),
+		Entry("hibernation.enabled = false and status.isHibernated is false", &core.Shoot{
+			Spec: core.ShootSpec{
+				Hibernation: &core.Hibernation{Enabled: &falseVar},
+			},
+		}, false),
+		Entry("hibernation.enabled = false and status.isHibernated is true", &core.Shoot{
+			Spec: core.ShootSpec{
+				Hibernation: &core.Hibernation{Enabled: &falseVar},
+			},
+			Status: core.ShootStatus{
+				IsHibernated: true,
+			},
+		}, true),
+		Entry("hibernation.enabled = true", &core.Shoot{
+			Spec: core.ShootSpec{
+				Hibernation: &core.Hibernation{Enabled: &trueVar},
+			},
+		}, true),
+	)
+
 	DescribeTable("#SeedSettingExcessCapacityReservationEnabled",
 		func(settings *core.SeedSettings, expectation bool) {
 			Expect(SeedSettingExcessCapacityReservationEnabled(settings)).To(Equal(expectation))

--- a/pkg/apis/core/validation/shoot.go
+++ b/pkg/apis/core/validation/shoot.go
@@ -1035,14 +1035,12 @@ func validateHibernationUpdate(new, old *core.Shoot) field.ErrorList {
 
 	if !hibernationEnabledInOld && hibernationEnabledInNew {
 		if new.Status.Credentials != nil && new.Status.Credentials.Rotation != nil && new.Status.Credentials.Rotation.ETCDEncryptionKey != nil {
-			etcdEncryptionKeyRotation := new.Status.Credentials.Rotation.ETCDEncryptionKey
-			if etcdEncryptionKeyRotation.Phase == core.RotationPreparing || etcdEncryptionKeyRotation.Phase == core.RotationCompleting {
+			if etcdEncryptionKeyRotation := new.Status.Credentials.Rotation.ETCDEncryptionKey; etcdEncryptionKeyRotation.Phase == core.RotationPreparing || etcdEncryptionKeyRotation.Phase == core.RotationCompleting {
 				allErrs = append(allErrs, field.Forbidden(fldPath, fmt.Sprintf("shoot cannot be hibernated when .status.credentials.rotation.etcdEncryptionKey.phase is %q", string(etcdEncryptionKeyRotation.Phase))))
 			}
 		}
 		if new.Status.Credentials != nil && new.Status.Credentials.Rotation != nil && new.Status.Credentials.Rotation.ServiceAccountKey != nil {
-			serviceAccountKeyRotation := new.Status.Credentials.Rotation.ServiceAccountKey
-			if serviceAccountKeyRotation.Phase == core.RotationPreparing || serviceAccountKeyRotation.Phase == core.RotationCompleting {
+			if serviceAccountKeyRotation := new.Status.Credentials.Rotation.ServiceAccountKey; serviceAccountKeyRotation.Phase == core.RotationPreparing || serviceAccountKeyRotation.Phase == core.RotationCompleting {
 				allErrs = append(allErrs, field.Forbidden(fldPath, fmt.Sprintf("shoot cannot be hibernated when .status.credentials.rotation.serviceAccountKey.phase is %q", string(serviceAccountKeyRotation.Phase))))
 			}
 		}

--- a/pkg/apis/core/validation/shoot.go
+++ b/pkg/apis/core/validation/shoot.go
@@ -151,6 +151,7 @@ func ValidateShootUpdate(newShoot, oldShoot *core.Shoot) field.ErrorList {
 	allErrs = append(allErrs, validateKubernetesVersionUpdate125(newShoot, oldShoot)...)
 	allErrs = append(allErrs, ValidateShoot(newShoot)...)
 	allErrs = append(allErrs, ValidateShootHAConfigUpdate(newShoot, oldShoot)...)
+	allErrs = append(allErrs, validateHibernationUpdate(newShoot, oldShoot)...)
 
 	return allErrs
 }
@@ -1022,6 +1023,32 @@ func isPSPDisabled(kubeAPIServerConfig *core.KubeAPIServerConfig) bool {
 		}
 	}
 	return false
+}
+
+func validateHibernationUpdate(new, old *core.Shoot) field.ErrorList {
+	var (
+		allErrs                 = field.ErrorList{}
+		fldPath                 = field.NewPath("spec", "hibernation", "enabled")
+		hibernationEnabledInOld = old.Spec.Hibernation != nil && pointer.BoolDeref(old.Spec.Hibernation.Enabled, false)
+		hibernationEnabledInNew = new.Spec.Hibernation != nil && pointer.BoolDeref(new.Spec.Hibernation.Enabled, false)
+	)
+
+	if !hibernationEnabledInOld && hibernationEnabledInNew {
+		if new.Status.Credentials != nil && new.Status.Credentials.Rotation != nil && new.Status.Credentials.Rotation.ETCDEncryptionKey != nil {
+			etcdEncryptionKeyRotation := new.Status.Credentials.Rotation.ETCDEncryptionKey
+			if etcdEncryptionKeyRotation.Phase == core.RotationPreparing || etcdEncryptionKeyRotation.Phase == core.RotationCompleting {
+				allErrs = append(allErrs, field.Forbidden(fldPath, fmt.Sprintf("shoot cannot be hibernated when .status.credentials.rotation.etcdEncryptionKey.phase is %q", string(etcdEncryptionKeyRotation.Phase))))
+			}
+		}
+		if new.Status.Credentials != nil && new.Status.Credentials.Rotation != nil && new.Status.Credentials.Rotation.ServiceAccountKey != nil {
+			serviceAccountKeyRotation := new.Status.Credentials.Rotation.ServiceAccountKey
+			if serviceAccountKeyRotation.Phase == core.RotationPreparing || serviceAccountKeyRotation.Phase == core.RotationCompleting {
+				allErrs = append(allErrs, field.Forbidden(fldPath, fmt.Sprintf("shoot cannot be hibernated when .status.credentials.rotation.serviceAccountKey.phase is %q", string(serviceAccountKeyRotation.Phase))))
+			}
+		}
+	}
+
+	return allErrs
 }
 
 // ValidateKubeAPIServer validates KubeAPIServerConfig.

--- a/pkg/apis/core/validation/shoot.go
+++ b/pkg/apis/core/validation/shoot.go
@@ -2087,8 +2087,8 @@ func validateShootOperation(operation, maintenanceOperation string, shoot *core.
 		if !availableShootOperations.Has(operation) {
 			allErrs = append(allErrs, field.NotSupported(fldPathOp, operation, sets.List(availableShootOperations)))
 		}
-		if helper.HibernationIsEnabled(shoot) && forbiddenShootOperationsWhenHibernated.Has(operation) {
-			allErrs = append(allErrs, field.Forbidden(fldPathOp, "operation is not permitted when shoot is hibernated"))
+		if helper.IsShootInHibernation(shoot) && forbiddenShootOperationsWhenHibernated.Has(operation) {
+			allErrs = append(allErrs, field.Forbidden(fldPathOp, "operation is not permitted when shoot is hibernated or is waking up"))
 		}
 	}
 
@@ -2096,8 +2096,8 @@ func validateShootOperation(operation, maintenanceOperation string, shoot *core.
 		if !availableShootMaintenanceOperations.Has(maintenanceOperation) {
 			allErrs = append(allErrs, field.NotSupported(fldPathMaintOp, maintenanceOperation, sets.List(availableShootMaintenanceOperations)))
 		}
-		if helper.HibernationIsEnabled(shoot) && forbiddenShootOperationsWhenHibernated.Has(maintenanceOperation) {
-			allErrs = append(allErrs, field.Forbidden(fldPathMaintOp, "operation is not permitted when shoot is hibernated"))
+		if helper.IsShootInHibernation(shoot) && forbiddenShootOperationsWhenHibernated.Has(maintenanceOperation) {
+			allErrs = append(allErrs, field.Forbidden(fldPathMaintOp, "operation is not permitted when shoot is hibernated or is waking up"))
 		}
 	}
 
@@ -2213,7 +2213,7 @@ func validateShootHAControlPlaneSpecUpdate(newShoot, oldShoot *core.Shoot, fldPa
 	if newShoot.Spec.ControlPlane != nil && newShoot.Spec.ControlPlane.HighAvailability != nil {
 		newVal = newShoot.Spec.ControlPlane.HighAvailability.FailureTolerance.Type
 		// TODO(@aaronfern): remove this validation of not allowing scale-up to HA while hibernated when https://github.com/gardener/etcd-druid/issues/589 is resolved
-		if !oldValExists && isShootInHibernation(newShoot) {
+		if !oldValExists && helper.IsShootInHibernation(newShoot) {
 			allErrs = append(allErrs, field.Forbidden(fldPath.Child("highAvailability", "failureTolerance", "type"), "Shoot is currently hibernated and cannot be scaled up to HA. Please make sure your cluster has woken up before scaling it up to HA"))
 		}
 	}
@@ -2237,12 +2237,4 @@ func isShootReadyForRotationStart(lastOperation *core.LastOperation) bool {
 		return true
 	}
 	return lastOperation.Type == core.LastOperationTypeReconcile
-}
-
-func isShootInHibernation(shoot *core.Shoot) bool {
-	if shoot.Spec.Hibernation != nil && shoot.Spec.Hibernation.Enabled != nil {
-		return *shoot.Spec.Hibernation.Enabled || shoot.Status.IsHibernated
-	}
-
-	return shoot.Status.IsHibernated
 }

--- a/pkg/apis/core/validation/shoot_test.go
+++ b/pkg/apis/core/validation/shoot_test.go
@@ -4468,6 +4468,64 @@ var _ = Describe("Shoot Validation Tests", func() {
 				Entry("rotate-serviceaccount-key-start", "rotate-serviceaccount-key-start"),
 				Entry("rotate-serviceaccount-key-complete", "rotate-serviceaccount-key-complete"),
 			)
+
+			DescribeTable("forbid hibernating the shoot when certain rotation operations are in progress",
+				func(status core.ShootStatus) {
+					shoot.Spec.Hibernation = &core.Hibernation{Enabled: pointer.Bool(true)}
+					shoot.Status = status
+
+					oldShoot := shoot.DeepCopy()
+					oldShoot.Spec.Hibernation = &core.Hibernation{Enabled: pointer.Bool(false)}
+
+					Expect(ValidateShootUpdate(shoot, oldShoot)).To(ContainElement(PointTo(MatchFields(IgnoreExtras, Fields{
+						"Type":  Equal(field.ErrorTypeForbidden),
+						"Field": Equal("spec.hibernation.enabled"),
+						"Detail": And(
+							ContainSubstring("shoot cannot be hibernated"),
+							Or(
+								ContainSubstring("phase is %q", "Preparing"),
+								ContainSubstring("phase is %q", "Completing"),
+							),
+						),
+					}))))
+				},
+				Entry("ETCD encryption key rotation is in Preparing phase", core.ShootStatus{
+					Credentials: &core.ShootCredentials{
+						Rotation: &core.ShootCredentialsRotation{
+							ETCDEncryptionKey: &core.ETCDEncryptionKeyRotation{
+								Phase: core.RotationPreparing,
+							},
+						},
+					},
+				}),
+				Entry("ETCD encryption key rotation is in Completing phase", core.ShootStatus{
+					Credentials: &core.ShootCredentials{
+						Rotation: &core.ShootCredentialsRotation{
+							ETCDEncryptionKey: &core.ETCDEncryptionKeyRotation{
+								Phase: core.RotationCompleting,
+							},
+						},
+					},
+				}),
+				Entry("ServiceAccount key rotation is in Preparing phase", core.ShootStatus{
+					Credentials: &core.ShootCredentials{
+						Rotation: &core.ShootCredentialsRotation{
+							ServiceAccountKey: &core.ServiceAccountKeyRotation{
+								Phase: core.RotationPreparing,
+							},
+						},
+					},
+				}),
+				Entry("ServiceAccount key rotation is in Completing phase", core.ShootStatus{
+					Credentials: &core.ShootCredentials{
+						Rotation: &core.ShootCredentialsRotation{
+							ServiceAccountKey: &core.ServiceAccountKeyRotation{
+								Phase: core.RotationCompleting,
+							},
+						},
+					},
+				}),
+			)
 		})
 	})
 

--- a/pkg/apis/core/validation/shoot_test.go
+++ b/pkg/apis/core/validation/shoot_test.go
@@ -4396,7 +4396,7 @@ var _ = Describe("Shoot Validation Tests", func() {
 					Expect(ValidateShoot(shoot)).To(ContainElement(PointTo(MatchFields(IgnoreExtras, Fields{
 						"Type":   Equal(field.ErrorTypeForbidden),
 						"Field":  Equal("metadata.annotations[gardener.cloud/operation]"),
-						"Detail": ContainSubstring("operation is not permitted when shoot is hibernated"),
+						"Detail": ContainSubstring("operation is not permitted when shoot is hibernated or is waking up"),
 					}))))
 					delete(shoot.Annotations, "gardener.cloud/operation")
 
@@ -4404,7 +4404,39 @@ var _ = Describe("Shoot Validation Tests", func() {
 					Expect(ValidateShoot(shoot)).To(ContainElement(PointTo(MatchFields(IgnoreExtras, Fields{
 						"Type":   Equal(field.ErrorTypeForbidden),
 						"Field":  Equal("metadata.annotations[maintenance.gardener.cloud/operation]"),
-						"Detail": ContainSubstring("operation is not permitted when shoot is hibernated"),
+						"Detail": ContainSubstring("operation is not permitted when shoot is hibernated or is waking up"),
+					}))))
+					delete(shoot.Annotations, "maintenance.gardener.cloud/operation")
+				},
+
+				Entry("rotate-credentials-start", "rotate-credentials-start"),
+				Entry("rotate-credentials-complete", "rotate-credentials-complete"),
+				Entry("rotate-etcd-encryption-key-start", "rotate-etcd-encryption-key-start"),
+				Entry("rotate-etcd-encryption-key-complete", "rotate-etcd-encryption-key-complete"),
+				Entry("rotate-serviceaccount-key-start", "rotate-serviceaccount-key-start"),
+				Entry("rotate-serviceaccount-key-complete", "rotate-serviceaccount-key-complete"),
+			)
+
+			DescribeTable("forbid certain rotation operations when shoot is waking up",
+				func(operation string) {
+					shoot.Spec.Hibernation = &core.Hibernation{Enabled: pointer.Bool(false)}
+					shoot.Status = core.ShootStatus{
+						IsHibernated: true,
+					}
+
+					metav1.SetMetaDataAnnotation(&shoot.ObjectMeta, "gardener.cloud/operation", operation)
+					Expect(ValidateShoot(shoot)).To(ContainElement(PointTo(MatchFields(IgnoreExtras, Fields{
+						"Type":   Equal(field.ErrorTypeForbidden),
+						"Field":  Equal("metadata.annotations[gardener.cloud/operation]"),
+						"Detail": ContainSubstring("operation is not permitted when shoot is hibernated or is waking up"),
+					}))))
+					delete(shoot.Annotations, "gardener.cloud/operation")
+
+					metav1.SetMetaDataAnnotation(&shoot.ObjectMeta, "maintenance.gardener.cloud/operation", operation)
+					Expect(ValidateShoot(shoot)).To(ContainElement(PointTo(MatchFields(IgnoreExtras, Fields{
+						"Type":   Equal(field.ErrorTypeForbidden),
+						"Field":  Equal("metadata.annotations[maintenance.gardener.cloud/operation]"),
+						"Detail": ContainSubstring("operation is not permitted when shoot is hibernated or is waking up"),
 					}))))
 					delete(shoot.Annotations, "maintenance.gardener.cloud/operation")
 				},

